extension/parquet/reader/variant/variant_binary_decoder.cpp

```diff
@@ -0,0 +1,360 @@
+	}
+	case VariantPrimitiveType::TIME_NTZ_MICROS: {
+		dtime_t micros_time;
+		micros_time.micros = Load<int64_t>(data);
+		auto value_str = Time::ToString(micros_time);
+		return yyjson_mut_strcpy(doc, value_str.c_str());
+	}
+	case VariantPrimitiveType::TIMESTAMP_NANOS: {
+		throw NotImplementedException("PrimitiveTypeDecode for VariantPrimitiveType::TIMESTAMP_NANOS");
+	}
+	case VariantPrimitiveType::TIMESTAMP_NTZ_NANOS: {
+		timestamp_ns_t nanos_ts;
+		nanos_ts.value = Load<int64_t>(data);
+
+		auto value = Value::TIMESTAMPNS(nanos_ts);
+		auto value_str = value.ToString();
+		return yyjson_mut_strcpy(doc, value_str.c_str());
+	}
+	case VariantPrimitiveType::UUID: {
+		auto uuid_value = UUIDValueConversion::ReadParquetUUID(data);
+		auto value_str = UUID::ToString(uuid_value);
+		return yyjson_mut_strcpy(doc, value_str.c_str());
+	}
+	default:
+		throw NotImplementedException("Variant PrimitiveTypeDecode not implemented for type (%d)",
+		                              static_cast<uint8_t>(value_metadata.primitive_type));
+	}
+}
+
+yyjson_mut_val *VariantBinaryDecoder::ShortStringDecode(yyjson_mut_doc *doc, const VariantMetadata &metadata,
+                                                        const VariantValueMetadata &value_metadata,
+                                                        const_data_ptr_t data) {
+	D_ASSERT(value_metadata.string_size < 64);
+	auto string_data = reinterpret_cast<const char *>(data);
+	if (!Utf8Proc::IsValid(string_data, value_metadata.string_size)) {
+		throw InternalException("Can't decode Variant short-string, string isn't valid UTF8");
+	}
+	return yyjson_mut_strncpy(doc, string_data, value_metadata.string_size);
+}
+
+yyjson_mut_val *VariantBinaryDecoder::ObjectDecode(yyjson_mut_doc *doc, const VariantMetadata &metadata,
+                                                   const VariantValueMetadata &value_metadata, const_data_ptr_t data) {
+	auto obj = yyjson_mut_obj(doc);
+
+	auto field_offset_size = value_metadata.field_offset_size;
+	auto field_id_size = value_metadata.field_id_size;
+	auto is_large = value_metadata.is_large;
+
+	idx_t num_elements;
+	if (is_large) {
+		num_elements = Load<uint32_t>(data);
+		data += sizeof(uint32_t);
+	} else {
+		num_elements = Load<uint8_t>(data);
+		data += sizeof(uint8_t);
+	}
+
+	auto field_ids = data;
+	auto field_offsets = data + (num_elements * field_id_size);
+	auto values = field_offsets + ((num_elements + 1) * field_offset_size);
+
+	idx_t last_offset = ReadVariableLengthLittleEndian(field_offset_size, field_offsets);
+	for (idx_t i = 0; i < num_elements; i++) {
+		auto field_id = ReadVariableLengthLittleEndian(field_id_size, field_ids);
+		auto next_offset = ReadVariableLengthLittleEndian(field_offset_size, field_offsets);
+
+		auto value = Decode(doc, metadata, values + last_offset);
+		auto &key = metadata.strings[field_id];
+		yyjson_mut_obj_add_val(doc, obj, key.c_str(), value);
+		last_offset = next_offset;
```

Comment:
Slight rant:
The spec says that an Object Variant should have `num_elements + 1` offsets, where the last offset points to the end of the `values`

This makes sense when the offsets can be used like: `offsets[i + 1] - offsets[i]` to get the length of the current `value`.
But this is not guaranteed, because the offsets aren't guaranteed to be monotically increasing:
> The field ids and field offsets must be in lexicographical order of the corresponding field names in the metadata dictionary. However, the actual value entries do not need to be in any particular order. This implies that the field_offset values may not be monotonically increasing